src/halo2/circuit_builder.rs

```diff
@@ -0,0 +1,625 @@
+use num_bigint::{BigInt, BigUint};
+use polyexen::expr::{ColumnKind, Expr, PlonkVar};
+use polyexen::plaf::backends::halo2::PlafH2Circuit;
+use polyexen::plaf::{ColumnFixed, ColumnWitness, Columns, Info, Lookup, Plaf, Poly, Witness};
+
+use crate::analyzer::{BinaryOperator, Expression, IdentityKind};
+use crate::number::AbstractNumberType;
+use crate::{analyzer, constant_evaluator, witness_generator};
+use num_traits::{One, Zero};
+
+use super::circuit_data::CircuitData;
+
+fn expression_2_expr(
+    cd: &CircuitData,
+    expr: &Expression,
+    int_to_field: &dyn Fn(&BigInt) -> BigUint,
+) -> (Expr<PlonkVar>, i32) {
+    match expr {
+        Expression::Number(n) => (Expr::Const(int_to_field(n)), 0),
+        Expression::PolynomialReference(polyref) => {
+            assert_eq!(polyref.index, None);
+
+            let rotation = if polyref.next { 1 } else { 0 };
+
+            let plonkvar = PlonkVar::ColumnQuery {
+                column: cd.col(&polyref.name),
+                rotation,
+            };
+
+            (Expr::Var(plonkvar), rotation)
+        }
+        Expression::BinaryOperation(lhe, op, rhe) => {
+            let (lhe, lhe_rot) = expression_2_expr(cd, lhe, int_to_field);
+            let (rhe, rhe_rot) = expression_2_expr(cd, rhe, int_to_field);
+            let res = match op {
+                BinaryOperator::Add => Expr::Sum(vec![lhe, rhe]),
+                BinaryOperator::Sub => Expr::Sum(vec![lhe, Expr::Neg(Box::new(rhe))]),
+                BinaryOperator::Mul => Expr::Mul(vec![lhe, rhe]),
+                _ => unimplemented!("{:?}", expr),
+            };
+            (res, std::cmp::max(lhe_rot, rhe_rot))
+        }
+
+        _ => unimplemented!("{:?}", expr),
+    }
+}
+
+pub(crate) fn analyzed_to_circuit(
+    analyzed: &analyzer::Analyzed,
+    query_callback: Option<impl FnMut(&str) -> Option<AbstractNumberType>>,
+    field_mod: BigUint,
+    verbose: bool,
+    int_to_field: &dyn Fn(&BigInt) -> BigUint,
+) -> PlafH2Circuit {
+    // The structure of the table is as following
+    //
+    // | constant columns | __enable_cur | __enable_next |  witness columns | \
+    // |  bla_bla_bla     |    1         |       1       |   bla_bla_bla    |  |
+    // |  bla_bla_bla     |    1         |       1       |   bla_bla_bla    |  |>  witness + fixed  2^(k-1)
+    // |  ...             |   ...        |      ...      |   ...            |  |
+    // |  bla_bla_bla     |    1         |       0       |   bla_bla_bla    | /     <- __enable_next == 0 since there's no state transition
+    // |  0               |    0         |       0       |   0              | \
```

Comment:
I think the lower half was removed.

Reply:
I thought you removed the code that creates the lower half, or did I misunderstand?